### PR TITLE
Test for Saml IdP Configuration

### PR DESCRIPTION
**Pull request by thomaschiozzi-tndigit**
Added a test for Saml IdP Configuration and Relying Party Registration.
*IMPORTANT*: also added a new dependancy in pox.xml

### CODE REVIEW

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+    /**
+     * This tests checks if s saml configuration map matches the content of the given saml configuration
+     * @throws Exception
+     */
+    @Test
+    public void validateSamlIdentityProviderConfigCreation() throws Exception {
+
+        // Create (hardcoded) configuration map to be used for initialization
+        Map<String, Serializable> rawCfgMap = buildRawConfigurationMap(true);
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:207]: given that this test **checks** the config and the map do not use the helper function, but build the map here and then check

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        SamlIdentityProviderConfigMap idProvCfgMap = new SamlIdentityProviderConfigMap();
+        idProvCfgMap.setConfiguration(rawCfgMap);
+
+        ConfigurableIdentityProvider cip = getDefaultConfigurableIdentityProvider();
+        IdentityProviderSettingsMap idProvSettingMap = getDefaultIdentityProviderSettingsMap();
+
+        SamlIdentityProviderConfig IdPCfg = new SamlIdentityProviderConfig(cip, idProvSettingMap, idProvCfgMap);
+        Assertions.assertThat(IdPCfg.getEntityId()).isEqualTo(expEntityID);
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:215]: add an assertion for every field you *explicitly* set in the map at 207

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    /**
+     * This test checks that a RelyingPartyRegistration contains information matching the saml configuration (with automatic idp metadata discovery)
+     * @throws Exception
+     */
+    @Test
+    public void validateRelyingPartyRegistrationAutomatic() throws Exception {
+
+        boolean automaticDiscoverySetting = true;
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:225]: drop

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        // Create (hardcoded) configuration map to be used for initialization
+        Map<String, Serializable> rawCfgMap = buildRawConfigurationMap(automaticDiscoverySetting);
+        SamlIdentityProviderConfigMap idProvCfgMap = new SamlIdentityProviderConfigMap();
+        idProvCfgMap.setConfiguration(rawCfgMap);
+
+        ConfigurableIdentityProvider cip = getDefaultConfigurableIdentityProvider();
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:231]: build these here

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        IdentityProviderSettingsMap idProvSettingMap = getDefaultIdentityProviderSettingsMap();
+
+        SamlIdentityProviderConfig idProvCfg = new SamlIdentityProviderConfig(cip, idProvSettingMap, idProvCfgMap);
+        // Run mock server to locally catch the identity provider metadata request.
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:235]: do this via the library, in the method you should just call the endpoint

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+        // check that the obtained saml Identity Provider matches the obtained map
+        RelyingPartyRegistration rpRegistration = idProvCfg.getRelyingPartyRegistration();
+        mockIdpServer.stop();
+
+        Assertions.assertThat(rpRegistration.getEntityId()).isEqualTo(expEntityID);
+        Assertions.assertThat(rpRegistration.getRegistrationId()).isEqualTo(expProviderId);
+        Assertions.assertThat(rpRegistration.getSigningX509Credentials().size()).isEqualTo(1);
+        for (Saml2X509Credential credential : rpRegistration.getSigningX509Credentials()) {
+            // NOTE: these assertions currently assume that there is only one x509credential(s) -> for-loop is trivial
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:248]: the test checks against a **known** configuration. If there is *one* credential you should check both:
- the credential is there and is the same
- there are NO other credentials

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+            // NOTE: if there is more than one x509credential(s), iteration order would NOT be guaranteed as we are iterating on a Set
+
+            // check private information
+            PrivateKey obtPrivateKey = credential.getPrivateKey();
+            String obtPrivKeyb64Enc = Base64.getEncoder().encodeToString(obtPrivateKey.getEncoded());
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:253]: this reasoning is flawed. Keys and certificates are stored with PEM, a specific format. Encoding the key back to base64 is *not* guaranteed to produce the same string. 
If you want to check the key you need to actually **parse** the key yourself and then check if it matches the one parsed by the class you are testing. Same for cert.

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+            String obtCertificateB64enc = Base64.getEncoder().encodeToString(obtCertificate.getEncoded());
+            Assertions.assertThat(obtCertificateB64enc).isEqualTo(expSigningCertificate);
+        }
+        Assertions.assertThat(rpRegistration.getAssertionConsumerServiceLocation()).isEqualTo(expRPRAssertionConsumerLocation);
+        Assertions.assertThat(rpRegistration.getAssertionConsumerServiceBinding()).isEqualTo(expRPRAssertionConsumerBinding);
+
+        RelyingPartyRegistration.AssertingPartyDetails partyDetails =  rpRegistration.getAssertingPartyDetails();
+        // sentinel value(s) used to check if auto configuration at least fetched data from the correct location
+        // This is a simplified approach. A deep and complete analysis will eventually be delegated to a second test
+        String expIdPEntityId = "https://idp.identityserver";
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:267]: where does this magic value come from? if its static set it in a constant

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    /**
+     * This test checks that a RelyingPartyRegistration contains information matching the saml configuration (with manual idp metadata discovery)
+     * @throws Exception
+     */
+    @Test
+    public void validateRelyingPartyRegistrationManual() throws Exception {
+
+        boolean automaticDiscoverySetting = false;
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:285]: drop this auto on/off flags

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+
+        // check that the obtained saml Identity Provider matches the obtained map
+        RelyingPartyRegistration rpRegistration = idProvCfg.getRelyingPartyRegistration();
+        Assertions.assertThat(rpRegistration.getEntityId()).isEqualTo(expEntityID);
+        Assertions.assertThat(rpRegistration.getRegistrationId()).isEqualTo(expProviderId);
+        Assertions.assertThat(rpRegistration.getSigningX509Credentials().size()).isEqualTo(1);
+        Assertions.assertThat(rpRegistration.getAssertionConsumerServiceLocation()).isEqualTo(expRPRAssertionConsumerLocation);
+        Assertions.assertThat(rpRegistration.getAssertionConsumerServiceBinding()).isEqualTo(expRPRAssertionConsumerBinding);
+
+        RelyingPartyRegistration.AssertingPartyDetails partyDetails =  rpRegistration.getAssertingPartyDetails();
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:304]: partyDetails could be null. Assert that is not the case before accessing properties in it

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+/**
+ * This test is aimed at validating if a configuration map will yield a valid
+ * and corresponding SamlProviderConfiguration.
+ */
+public class SamlIdentityProviderConfigurationTest {
+
+    private static final int IDP_METADATA_MOCK_PORT = 51994;
+    private static final String MOCK_METADATA_ENDPOINT = "metadata";
+    private static final String IDP_METADATA_EXAMPLE_FP = "saml/saml-idp-metadata-example.xml";
+    private static final String expProviderId = "ProviderTestId";
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:32]: what is the `exp` prefix on every variable?

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+        // Run mock server to locally catch the identity provider metadata request.
+        String IdpMetadata = fetchIdpMetadataExample();
+        WireMockServer mockIdpServer = buildMockIdpServer(IdpMetadata);
+        mockIdpServer.start();
+
+        // check that the obtained saml Identity Provider matches the obtained map
+        RelyingPartyRegistration rpRegistration = idProvCfg.getRelyingPartyRegistration();
+        mockIdpServer.stop();
+
+        Assertions.assertThat(rpRegistration.getEntityId()).isEqualTo(expEntityID);
```
Comment (matteo-s) [inline src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java:244]: static import assertThat etc
```
import static org.assertj.core.api.Assertions.assertThat;
```

File: src/test/resources/saml/saml-idp-metadata-example.xml [hunk 13]
```diff
@@ -0,0 +1,48 @@
+<EntityDescriptor
+        ID="_c066524f-ba36-49d5-9dfa-ae14e13c1392"
+        entityID="https://idp.identityserver"
+        validUntil="2022-07-20T09:48:54Z"
```
Comment (matteo-s) [inline src/test/resources/saml/saml-idp-metadata-example.xml:4]: drop the date. Besides it is *already* expired..

File: src/test/resources/saml/saml-idp-metadata-example.xml [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        cacheDuration="PT15M"
+        xmlns="urn:oasis:names:tc:SAML:2.0:metadata"
+        xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">
+
+    <IDPSSODescriptor WantAuthnRequestsSigned="true" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
+        <SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://idp.identityserver/saml/sso" />
+        <SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://idp.identityserver/saml/sso" />
```
Comment (matteo-s) [inline src/test/resources/saml/saml-idp-metadata-example.xml:11]: drop everything except  HTTP-REDIRECT, we support only this flow

File: src/test/resources/saml/saml-idp-metadata-example.xml [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        <SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact" Location="https://idp.identityserver/saml/sso" />
+
+        <SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://idp.identityserver/saml/slo" />
+        <SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://idp.identityserver/saml/slo" />
+        <SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact" Location="https://idp.identityserver/saml/slo" />
+
+        <ArtifactResolutionService Binding="urn:oasis:names:tc:SAML:2.0:bindings:SOAP" Location="https://idp.identityserver/saml/ars" index="0" />
+
+        <NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</NameIDFormat>
```
Comment (matteo-s) [inline src/test/resources/saml/saml-idp-metadata-example.xml:20]: We do not support SAML 1.1. Drop everything related

File: src/test/resources/saml/saml-idp-metadata-example.xml [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+                </X509Data>
+            </KeyInfo>
+        </KeyDescriptor>
+    </IDPSSODescriptor>
+
+    <Organization>
```
Comment (matteo-s) [inline src/test/resources/saml/saml-idp-metadata-example.xml:35]: drop everything that is optional: organization, contact etc

File: src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java (hunk 7 above)
Comment (thomaschiozzi-tndigit) [reply src/test/java/it/smartcommunitylab/aac/saml/SamlIdentityProviderConfigurationTest.java]: Parsed the key using [the linked spiring test](https://github.com/spring-projects/spring-security/blob/main/saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/registration/OpenSamlMetadataRelyingPartyRegistrationConverterTests.java) as reference

File: pom.xml [hunk 17]
```diff
@@ -342,6 +355,12 @@
 			<artifactId>webauthn-server-core</artifactId>
 			<version>${yubico.version}</version>
 		</dependency>
+		<dependency>
```
Comment (matteo-s) [inline pom.xml:358]: do you really need this additional dependency?

Comment (thomaschiozzi-tndigit) [reply pom.xml:358]: No, the dependency is not required and was a leftover from previous version of the code. To be removed in next commit.

**Pull request merged**